src/test/java/io/debezium/connector/spanner/LowWatermarkRecordIT.java

```diff
@@ -0,0 +1,86 @@
+ */
+package io.debezium.connector.spanner;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.Assert.assertTrue;
+
+import java.time.Instant;
+import java.time.format.DateTimeFormatter;
+import java.time.temporal.ChronoUnit;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
+
+import org.apache.kafka.connect.data.Struct;
+import org.apache.kafka.connect.source.SourceRecord;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+import io.debezium.config.Configuration;
+
+public class LowWatermarkRecordIT extends AbstractSpannerConnectorIT {
+
+    private static final String tableName = "low_watermark_record_tests_table";
+    private static final String changeStreamName = "lowWatermarkRecordTestChangeStream";
+
+    @BeforeAll
+    static void setup() throws InterruptedException, ExecutionException {
+        databaseConnection.createTable(tableName + "(id int64, name string(100)) primary key(id)");
+        databaseConnection.createChangeStream(changeStreamName, tableName);
+
+        System.out.println("LowWatermarkRecordIT is ready...");
+    }
+
+    @AfterAll
+    static void clear() throws InterruptedException {
+        databaseConnection.dropChangeStream(changeStreamName);
+        databaseConnection.dropTable(tableName);
+    }
+
+    @Test
+    public void shouldStreamUpdatesToKafka() throws InterruptedException {
+        Instant now = Instant.now();
+        final Configuration config = Configuration.copy(baseConfig)
+                .with("gcp.spanner.change.stream", changeStreamName)
+                .with("name", tableName + "_test")
+                .with("gcp.spanner.start.time",
+                        DateTimeFormatter.ISO_INSTANT.format(now))
+                .with("gcp.spanner.low-watermark.enabled", true)
+                .build();
+        initializeConnectorTestFramework();
+        start(SpannerConnector.class, config);
+        assertConnectorIsRunning();
+        databaseConnection.executeUpdate("insert into " + tableName + "(id, name) values (1, 'some name')");
+        databaseConnection.executeUpdate("update " + tableName + " set name = 'test' where id = 1");
+        waitForCDC();
+        SourceRecords sourceRecords = consumeRecordsByTopic(10, false);
+        List<SourceRecord> records = sourceRecords.recordsForTopic(getTopicName(config, tableName));
+        List<Long> lowWatermarks = records.stream()
+                .map(rec -> rec.value() != null
+                        ? (Long) ((Struct) ((Struct) rec.value()).get("source")).get("low_watermark")
+                        : null)
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+        assertThat(!lowWatermarks.isEmpty());
```

Comment:
Is it possible to change the test such that we wait until the connector's low watermark passes T+5 seconds? And then we cancel the connector?

Reply:
The reason not to write the test in this long-pulling mechanism is due to the limitation of `consumeRecordsByTopic(int numOfRecords)`. This test util function pulls the first n records from all topics. Due to the more frequent heartbeat records(shorter churning interval). If we put this test logic in a while loop, it might look like:
```
int initialPullingNum = 10;
startConnector();
while(watermark<queryStart+5Seconds){
    watermark = largest watermark among consumeRecordsByTopic(initialPullingNum+10);
}
stopConnector();
```
Which means we can't read from the record index from last iteration, and we need to start from fresh and pulling more records than previous iteration. This might take quite a long time.

In fact, there is a perfect test util function for this test, which is [[consumeRecordsUntil](https://github.com/debezium/debezium/blob/84ab7f7b8adc9b9c437ea5c5829ee83b86d3a761/debezium-embedded/src/test/java/io/debezium/embedded/AbstractConnectorTest.java#L675), which we can easily pass in a predicate like `watermark>start+5Seconds`. However this functions requires the records to pass the record serialization verification(the record formats for all connectors under debezium main repo, while our connector does not follow it). Thus I thought just verify a record has watermark > 2 seconds should be good for a proof of functionality.

Reply:
Sure, sounds fair.